### PR TITLE
Add a rundk tool to help reproduce the same test flows as run in CI

### PR DESCRIPTION
**Pull request by chizhg**
**What this PR does, why we need it**:
1. Add a `rundk` tool to help reproduce the same test flows as run in CI
2. Update it to use `CopyAndAddMount` instead of `mount` to make the tool also being able to run on MacOS

**Which issue(s) this PR fixes**:
#2063

### CODE REVIEW

**Comment by chizhg** [comment]
/cc @coryrc @albertomilan 
FYI @zhongduo

**Comment by zhongduo** [comment]
IIUC, this is trying to add a wrapper of the `docker run` command. But from my experience, the main complexity seems in the project setup and service account creation and role binding, as well as the `IS_PROW` check. Is there any plan to make that easier too?

**Comment by chizhg** [comment]
> IIUC, this is trying to add a wrapper of the `docker run` command. But from my experience, the main complexity seems in the project setup and service account creation and role binding, as well as the `IS_PROW` check. Is there any plan to make that easier too?

For the project setup and service account creation and role binding, https://github.com/knative/test-infra/blob/master/config/prod/build-cluster/boskos/set_up_boskos_project.sh is the script we run for our CI and it incorporates everything for all the Knative projects. For each project, they should maintain their own docs for the related setups, e.g. https://github.com/google/knative-gcp/blob/master/test/e2e/README.md. The doc generally looks good to me but some improvements can be made there, e.g. you can update this doc to add instructions on how to run the test flow with this `rundk` tool after this PR is merged.

Also test-infra only exposes `IS_PROW` as an env vars, and it's the responsibility of each Knative project to decide how to use it. So IMO if you find `IS_PROW` check as a problem, the issue should be fixed by the project owner, in this case is `knative-gcp`.

**Comment by sethvargo** [comment]
I've been able to get away with a simple shell script most of the time with our use of prow. Example:

```sh
$ docker run 

# interactive mode (so we get a tty)
-it 

# remove it when were done
--rm

# mount the source code in /src (that's where prow wants them)
-v ~/Development/knative:/src

# mount go module cache - it'll work without this, but it will download all
# dependencies each time, which is annoying
-v ~/Development/go/pkg:/go/pkg

# mount docker on the host to docker in the container - required for DinD
-v $(which docker):/usr/bin/docker
-v /var/run/docker.sock:/var/run/docker.sock

# on a mac, set to host networking to make DinD work. not required on linux
--network host

# mount gcloud credentials to get default app creds, readonly to be safe
-v ~/.config/gcloud:/root/.config/gcloud:ro

# set workdir (this might be the default, didn't test)
--workdir /src

# use our runner
my.image.url

# execute tests
runtests.sh
```

**Comment by chizhg** [comment]
> I've been able to get away with a simple shell script most of the time with our use of prow. Example:
> 
> ```shell
> $ docker run 
> 
> # interactive mode (so we get a tty)
> -it 
> 
> # remove it when were done
> --rm
> 
> # mount the source code in /src (that's where prow wants them)
> -v ~/Development/knative:/src
> 
> # mount go module cache - it'll work without this, but it will download all
> # dependencies each time, which is annoying
> -v ~/Development/go/pkg:/go/pkg
> 
> # mount docker on the host to docker in the container - required for DinD
> -v $(which docker):/usr/bin/docker
> -v /var/run/docker.sock:/var/run/docker.sock
> 
> # on a mac, set to host networking to make DinD work. not required on linux
> --network host
> 
> # mount gcloud credentials to get default app creds, readonly to be safe
> -v ~/.config/gcloud:/root/.config/gcloud:ro
> 
> # set workdir (this might be the default, didn't test)
> --workdir /src
> 
> # use our runner
> my.image.url
> 
> # execute tests
> runtests.sh
> ```

Thanks @sethvargo ! Yes, the underlying implementation of this tool is basically a wrapper of a `docker` command, but it's more generic, more similar as run on Prow and easier to use (e.g. it can be run under any subdirectory of a Knative repo and it will run the test flow from the root folder of that repo).

File: pkg/interactive/docker.go
Comment (albertomilan) [inline pkg/interactive/docker.go:90]: Why not `0770`, given it's a directory?

File: pkg/interactive/docker.go
Comment (albertomilan) [inline pkg/interactive/docker.go:94]: Can we have some error checking control done here with a log statement?

File: pkg/interactive/docker.go
Comment (albertomilan) [reply pkg/interactive/docker.go:90]: I meant `0666`

File: pkg/interactive/command.go
Comment (coryrc) [inline pkg/interactive/command.go:17]: No, half the library is for running any commands from Go

File: rundk/main.go
Comment (albertomilan) [inline rundk/main.go:129]: If this is an important step, I would suggest to expand this to 5+ seconds. Reading the instruction will take ~1 sec already and we need to give room for the reaction from the user

File: pkg/interactive/docker.go
Comment (coryrc) [inline pkg/interactive/docker.go:84]: I need this function, don't remove it

Comment (coryrc) [reply pkg/interactive/docker.go:84]: You can add a new one if you want

File: pkg/interactive/docker.go
Comment (coryrc) [inline pkg/interactive/docker.go:83]: The purpose of this library is to write interactive scripts, not programs which can recover from errors and do smart things. Adding error checking makes the scripts much less focused on the task at hand and hard to parse.

If you want different functionality, put it in a different library

**Review by coryrc — CHANGES_REQUESTED** [review]

**Comment by coryrc** [comment]
/hold

Can't delete AddRWOverlay

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go]: Changed to `0666`.

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go]: Added some checking control here, but since it'll be a best-effort cleanup so its failure should not impact the final results, I'm using `warning` instead of `fatalf` log here.

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:129]: It's not a very important step, and also since users could be running this tool again and again, I feel 5+ seconds is a bit too long, so I'm inclined to keep it as 3 seconds.

File: pkg/interactive/command.go
Comment (chizhg) [reply pkg/interactive/command.go]: I've reverted the change here, but it's probably better to put these functions under different packages..

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go]: I have got the function back.

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go:85]: Hmm I'm not quite sure what makes the difference between them, and we already have another function that returns error instead of fatal - https://github.com/knative/test-infra/blob/master/pkg/interactive/docker.go#L43-L56

File: pkg/interactive/docker.go
Comment (coryrc) [reply pkg/interactive/docker.go]: um, directories need to be executable and we're writing to it, so why isn't it `0777`?

File: pkg/interactive/docker.go
Comment (coryrc) [reply pkg/interactive/docker.go:85]: Because that error is optional -- sometimes it's okay to fail and sometimes it is not.

If the overlay fails, it's unlikely you want to continue

File: rundk/main.go
Comment (coryrc) [reply rundk/main.go:129]: The purpose is: what if you just want the env setup and to use it yourself; then 3 seconds is fast enough because you'll be watching it.
Then why not wait on user input? I hate when you want to run something which takes a while, you launch it, walk away, then it doesn't start because of unnecessary user input.

File: pkg/interactive/docker.go
Comment (albertomilan) [reply pkg/interactive/docker.go]: Cory is correct. Executable flag for directories grant permission on its content. Discard this comment.

Comment (chizhg) [reply pkg/interactive/docker.go:92]: Yeah I just tried `0666` and it will actually throw out an error. I've reverted the change.

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go]: I have change it to fatal instead of returning an error.

File: rundk/README.md
Comment (coryrc) [inline rundk/README.md:41]: I don't think you should take the script name like this, instead of an argument, because if you run it from a subdirectory and use a relative path, it won't work. It also won't work with relative paths.

File: rundk/README.md
Comment (coryrc) [inline rundk/README.md:27]: The likelihood the local directory has the same path as would be useful inside the image is unlikely, so I don't think this feature adds anything

File: rundk/README.md
Comment (coryrc) [inline rundk/README.md:29]: These two options don't mention that the env vars get promoted to the image

File: rundk/README.md
Comment (coryrc) [reply rundk/README.md:41]: So maybe take path -> convert to realpath -> confirm under the repo root -> convert to path relative to repo root?

**Review by coryrc — CHANGES_REQUESTED** [review]

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:41]: It will work, since in this tool we set the working directory as the root directory of the repo, as long as you run this tool in one of the repo's subdirectories.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:27]: Users would probably want to mount other files when they run the tests? The flag is empty by default, just an extension in case users may want extra mounts.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:29]: Updated the README here.

File: rundk/README.md
Comment (coryrc) [reply rundk/README.md:42]: if I'm in the test directory already, and run:
`rundk ./e2e-tests.sh`
it won't work despite auto-completing that way

File: rundk/README.md
Comment (coryrc) [reply rundk/README.md:27]: What files? The whole git directory gets mounted.
What good does mounting `/home/<myusername>/.kube/config` do, it's not put in a usable place?
You need to be able to define the external and internal paths separately to be useful.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:43]: Good point. I have changed the working directory to be the directory where the tool is run.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:27]: Changed the flag value to be in the format of ``source1:target1,source2:target2`.
Thanks.

File: rundk/main.go
Comment (albertomilan) [reply rundk/main.go:130]: I do agree 3 seconds might be enough for a user who is used to invoking the tool, but it might leave ~1-2 seconds of reaction to a developer who does read the message for the first time.
In any case, being the outcome controlled I'm fine with either option.

+1 to @coryrc 's comment. Waiting on user's input does not seem to be the best option at all.

File: rundk/main.go
Comment (coryrc) [inline rundk/main.go:87]: You should use a constant filename for the file inside the image

File: rundk/main.go
Comment (coryrc) [inline rundk/main.go:107]: ditto filename

File: rundk/main.go
Comment (coryrc) [inline rundk/main.go:144]: you don't want to mount the repoRoot to `/go/src/...`?

File: pkg/interactive/docker.go
Comment (coryrc) [inline pkg/interactive/docker.go:113]: why bash? you should be able to run `cp` directly

File: rundk/README.md
Comment (coryrc) [inline rundk/README.md:35]: why is this double `--` but the others are single `-` prefix?

File: rundk/README.md
Comment (coryrc) [reply rundk/README.md:41]: I don't see this change?

File: rundk/README.md
Comment (coryrc) [inline rundk/README.md:42]: Bad example because there's special handling, `KO_DOCKER_REPO` would be a better example

**Review by coryrc — CHANGES_REQUESTED** [review]

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:144]: With Go mod, it does not really matter where we put the repo, it'll work under any folder.

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go]: The reason was I used to run as `cp -r xxx/* yyy`, and `*` can only be used with bash. But now since I'm not using `*` anymore, I'll remove the bash part then.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:35]: Sorry, copy&paste error. Fixed.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md]: It's now empty by default, so I've removed the whole `(default "GOOGLE_APPLICATION_CREDENTIALS")` part.

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:87]: Hmm do you mean always mounting the key file to the same target path? If yes are there any reasons or advantages?

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:107]: Similar as the above comment related to `GOOGLE_APPLICATION_CREDENTIALS`. Also the `KUBECONFIG` can be pointed to multiple kubeconfig files (for multiple clusters), so it's not quite easy to use constant filenames here.

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:144]: And this tool will need to work with other test images, which do not necessarily have their `GOPATH` set to `/go`.

File: rundk/main.go
Comment (coryrc) [reply rundk/main.go:87]: Yes:
Privacy: It leaks information about the caller. (it's mounting their home directory setup, not the biggest deal in the world)
Debugging: Easier to know the file will be in a fixed position (Chi: running this doesn't work for me; Cory: can you `cat /gac.json`?)
Defense-in-depth: Weird file paths could break things in scripts, setting to a fixed, simple name avoids that

File: rundk/README.md
Comment (coryrc) [inline rundk/README.md:42]: You are missing the important text that said environment variables will be promoted to the container

File: test/e2e-tests.sh
Comment (coryrc) [inline test/e2e-tests.sh:31]: I don't think we should hardcode your personal project

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:87]: 1. I don't think it will leak information about the caller, they will run the docker command and start the container on their local anyway..
2. If `GOOGLE_APPLICATION_CREDENTIALS` is set on the local, it must be explicitly set by the user, so they must know where the file is, and I don't think it's harder to find than asking us the magic value we hardcoded.
3. I don't seem to realize there could be weird file paths that could break things, especially since we only mount it as a single file. Do you have a concrete example or have you met issues related to that?

File: test/e2e-tests.sh
Comment (chizhg) [reply test/e2e-tests.sh]: Sorry this file was for my local testing and I committed it by mistake. Just deleted it.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md]: Added.

File: pkg/interactive/docker.go
Comment (albertomilan) [inline pkg/interactive/docker.go:102]: Being this a directory removal, why not making it an independent function and simply return the directory path as end result?
I'm not sure if this is a common practice in Go, but I'm not able to see the direct benefits of dealing with the function as a return object (it mainly complicates its interface)

File: pkg/interactive/docker.go
Comment (albertomilan) [inline pkg/interactive/docker.go:111]: `source += string(os.PathSeparator)+"."` ?

File: rundk/README.md
Comment (albertomilan) [inline rundk/README.md:30]: Nit: Typo in `fllow`

File: rundk/README.md
Comment (albertomilan) [inline rundk/README.md:44]: From a use perspective, it's not clear what might make an env var optional or not, given in both cases they will be promoted to the container.

File: rundk/main.go
Comment (albertomilan) [inline rundk/main.go:66]: Is the template `prow-docker.` working as expected? From the documentation [1], it seems `*` should be used to avoid overwriting.

[1] https://golang.org/pkg/io/ioutil/#TempDir

File: rundk/main.go
Comment (albertomilan) [inline rundk/main.go:190]: Nit: Use a unique reference for the wait time. It will simplify any potential future change which can diverge these values

Comment (chizhg) [reply rundk/main.go]: Done.

File: rundk/main.go
Comment (chizhg) [reply rundk/main.go:70]: The doc also mentioned `The directory name is generated by taking pattern and applying a random string to the end.`, so if it does not have `*`, it will simply do appending instead of replacing so we will get `prow-docker.xyz`. I have verified it on my local.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md:44]: This is what we have for the internal version of this tool that was implemented by Cory, and we do have reasons to keep it I think.

File: rundk/README.md
Comment (chizhg) [reply rundk/README.md]: Fixed, thanks.

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go]: Fixed.

File: pkg/interactive/docker.go
Comment (chizhg) [reply pkg/interactive/docker.go:102]: It's a common pattern in Go: a function returns another function that does the cleanup, and the callsite call `defer cleanup()` to make sure it's always being called, https://github.com/knative/serving/blob/master/cmd/queue/execprobe.go#L109 is one example and there are a lot more examples in other places.

File: rundk/README.md
Comment (albertomilan) [reply rundk/README.md:44]: Could you briefly describe the difference in the documentation and here so developers can make correct use of it?
I was not worried about the change, but they potential interpretation given the lack of explanation.

Comment (chizhg) [reply rundk/README.md]: Done. Please let me know if it looks reasonable to you now.

**Pull request merged**
